[#102] improvement: improve startup script

orenccl opened the pull request:
### What changes were proposed in this pull request?

1. Removed the debug option `set -x`.
2. Fixed warnings identified by IntelliJ, such as changing `cd "${playground_dir}" >/dev/null` to `cd "${playground_dir}" >/dev/null || exit 1` to prevent failures when `cd` fails.
3. Updated `docker run --pull always hello-world >/dev/null 2>&1` to `docker run --rm --pull always hello-world:latest >/dev/null 2>&1` to avoid pulling multiple images and ensure the container is removed after it finishes.
4. Changed naming conventions and log formats for consistency.
5. Added disk, RAM, and CPU checks.
6. Added the `--skip-checks` option for a quicker startup.
7. Corrected the log suffix format from `%Y%m%d%H%m%s` to `%Y%m%d%H%M%s`.
8. Removed the unnecessary `Confirm the requirement is available in your OS [Y/n]:`.
9. Combined port check status into one line instead of multiple lines.

### Why are the changes needed?

To improve the user experience.
fix #102
fix #113 

### Does this PR introduce _any_ user-facing change?

Added a new option `--skip-checks|-s` for a quicker startup.

### How was this patch tested?

Tested with the following commands:
- `./playground start`
- `./playground status`
- `./playground stop`
- `./playground start --skip-checks`
- `./playground start -s`

These tests were run to ensure the script works as intended.

![image](https://github.com/user-attachments/assets/6a5722a8-5860-4d24-8ac0-b8f66dabf041)

orenccl commented:
Hi @unknowntpo,

I've resolved the conflicts based on your recently merged PR #56.

Would you please review this? If possible, I would appreciate if you could also test it in your environment.

Thanks!

![image](https://github.com/user-attachments/assets/d3d42c57-1f3f-4ef3-84db-2d67f1964fef)

xunliu commented on playground.sh:
```diff
@@ -18,44 +18,104 @@
-  docker run --pull always hello-world >/dev/null 2>&1
+  echo "[INFO] Testing Docker environment by running hello-world..."
+  # Use always to test network connection
+  docker run --rm --pull always hello-world:latest >/dev/null 2>&1
```
Please specify `hello-world` fixed tag, do not use `latest`

orenccl replied:
I use the latest tag because `docker --pull always` ensures that an image is always downloaded from the network and updates the local image. Therefore, using any specific tag would yield the same result. However, I can still switch to a more stable version for consistency.

orenccl replied:
Got it.

xunliu commented on playground.sh:
```diff
@@ -77,50 +137,71 @@ checkHelm() {
+      checkDockerDisk
+      checkDockerRam
+      checkDockerCpu
+
```
Please remove this blank line.

xunliu commented on playground.sh:
```diff
@@ -152,16 +233,16 @@ status() {
   case "$runtime" in
   k8s)
-	helm uninstall --namespace gravitino-playground gravitino-playground 
+  helm uninstall --namespace gravitino-playground gravitino-playground
```
I think need to add spaces in the line head, keep code sytle.

xunliu commented on playground.sh:
```diff
@@ -177,27 +258,22 @@ docker)
+  echo "[ERROR] please specify which runtime you want to use, available runtime: [docker|k8s]"
+esac
+
+skipChecks=false
+
+case "$3" in
+--skip-checks)
+  skipChecks=true;
+  ;;
+-s)
+  skipChecks=true;
+  ;;
```
Do we need `skipChecks` startup param? It feels too complicated.
I think we must check docker or k8s envorinments.

orenccl replied:
Thank you for the feedback. I'll remove the skipChecks parameter as suggested. The test didn't take much time, so this change should be straightforward.

xunliu commented:
hi @orenccl Thank you for your improve playground
I think we didn't neet specify `docker|k8s` param, we can check use environment, 
1. if user only have docker, then launch playground on Docker
2. if use only have k8s, then launch playground on Docker
3. if use both have docker and k8s, then output a menu, let use select one to launch.

orenccl commented:
![image](https://github.com/user-attachments/assets/9738cf4a-4552-43d7-b4a7-999db4e5d830)

![image](https://github.com/user-attachments/assets/17029e96-89ab-45d5-b6ff-67654cc6f354)

![image](https://github.com/user-attachments/assets/ef7c942e-2c8a-4796-893d-8e693dc97074)

unknowntpo commented:
> * if user only have docker, then launch playground on Docker
> 
> * if use only have k8s, then launch playground on Docker
> 
> * if use both have docker and k8s, then output a menu, let use select one to launch.

@xunliu I think letting user to run both k8s and docker is more handy.

xunliu commented on playground.sh:
```diff
@@ -18,44 +18,107 @@
+
+testK8s() {
+  echo "[INFO] Testing K8s environment..."
+  kubectl cluster-info
```
I think we need to check if there exists `kubectl` first.

orenccl replied:
Before the `start`, `status`, or `stop` commands are run, the `checkCurrentRuntime` function will be executed.

The `checkCurrentRuntime` function checks whether `kubectl` is present.

![image](https://github.com/user-attachments/assets/5082adde-a885-4f99-b7ce-38babf2baef5)

![image](https://github.com/user-attachments/assets/2310dbb8-6bb1-4787-8320-7954d0676443)

xunliu commented on playground.sh:
```diff
@@ -64,140 +127,277 @@ checkHelm() {
   # version will be like:
   # Version:"v3.15.2"
   regex="Version:\"(v[0-9]\.[0-9]+\.[0-9])\""
   version=$(helm version)
```
I think we needs check if exist `helm` first.

orenccl replied:
This section checks if Helm exists:

```bash
isExist=$(command -v helm)
if [ "$isExist" ]; then
```

I've unified the usage of `which` and `command -v` to use `command -v` for consistent style.

```bash
if command -v helm >/dev/null 2>&1; then
```

xunliu commented on playground.sh:
```diff
@@ -64,140 +127,277 @@ checkHelm() {
     ;;
   docker)
-    logSuffix=$(date +%Y%m%d%H%m%s)
-    docker-compose up --detach
```
`docker-compose` or `docker compose`？
My macOS only have `docker-compose`.
`docker compose` is Linux OS?
Please check it.

Maybe you needs to check operation OS, and add different process.

orenccl replied:
Based on [this documentation](https://docs.docker.com/compose/releases/migrate/), I use `docker-compose`, which works on every operating system.

orenccl replied:
It's not about the OS; it's about the Docker version. Docker version 2 has a built-in `docker compose` command, whereas Docker version 1 does not. For Docker version 1, you need to use `docker-compose`, which is now deprecated.

I added a function to automatically detect the appropriate Docker Compose command.

**Reference**:
[Docker Compose Migration Guide](https://docs.docker.com/compose/releases/migrate/)

orenccl commented:
Update this PR,
1. [Unified port number for Docker and Kubernetes.](https://github.com/apache/gravitino-playground/pull/103/commits/4e8b0607cf0490493c3839694f255daec75e0ff6)
2. [update readme](https://github.com/apache/gravitino-playground/pull/103/commits/e9bca19713063ed86f990ab327a26d313f0bf902)

jerryshao commented:
Hi @orenccl would you please rebase this PR to fix the conflicts?

danhuawang commented:
> Update this PR:
> 
> 1. [Unified port number for Docker and Kubernetes.](https://github.com/apache/gravitino-playground/pull/103/commits/4e8b0607cf0490493c3839694f255daec75e0ff6)
> 2. [update readme](https://github.com/apache/gravitino-playground/pull/103/commits/e9bca19713063ed86f990ab327a26d313f0bf902)
> 
> Please let me know if there are any problems.

@orenccl Can you help check this issue when I ran the fileset example in jupyter? By the way ,we have 4 examples in jupyter, can you help verify them by access http://localhost:18888 ,especially k8s runtime.

<img width="1476" alt="image" src="https://github.com/user-attachments/assets/3007adee-8ac1-4418-913a-94b7f9d20c01">
```
---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
Cell In[39], line 12
     10 # Create metalake via Gravitino admin client
     11 metalake_name="default"
---> 12 metalake = gravitino_admin_client.create_metalake(name=metalake_name,
     13                                                   comment="metalake comment", 
     14                                                   properties={})
     15 print(metalake)

File /opt/conda/lib/python3.11/site-packages/gravitino/client/gravitino_admin_client.py:80, in GravitinoAdminClient.create_metalake(self, name, comment, properties)
     77 req = MetalakeCreateRequest(name, comment, properties)
     78 req.validate()
---> 80 resp = self._rest_client.post(
     81     self.API_METALAKES_LIST_PATH, req, error_handler=METALAKE_ERROR_HANDLER
     82 )
     83 metalake_response = MetalakeResponse.from_json(resp.body, infer_missing=True)
     84 metalake_response.validate()

File /opt/conda/lib/python3.11/site-packages/gravitino/utils/http_client.py:229, in HTTPClient.post(self, endpoint, json, error_handler, **kwargs)
    228 def post(self, endpoint, json=None, error_handler=None, **kwargs):
--> 229     return self._request(
    230         "post", endpoint, json=json, error_handler=error_handler, **kwargs
    231     )

File /opt/conda/lib/python3.11/site-packages/gravitino/utils/http_client.py:202, in HTTPClient._request(self, method, endpoint, params, json, data, headers, timeout, error_handler)
    197     request.add_header(
    198         AuthConstants.HTTP_HEADER_AUTHORIZATION,
    199         self.auth_data_provider.get_token_data().decode("utf-8"),
    200     )
    201 request.get_method = lambda: method
--> 202 is_success, resp = self._make_request(opener, request, timeout=timeout)
    204 if is_success:
    205     return resp

File /opt/conda/lib/python3.11/site-packages/gravitino/utils/http_client.py:139, in HTTPClient._make_request(self, opener, request, timeout)
    137 timeout = timeout or self.timeout
    138 try:
--> 139     return (True, Response(opener.open(request, timeout=timeout)))
    140 except HTTPError as err:
    141     err_body = err.read()

File /opt/conda/lib/python3.11/urllib/request.py:519, in OpenerDirector.open(self, fullurl, data, timeout)
    516     req = meth(req)
    518 sys.audit('urllib.Request', req.full_url, req.data, req.headers, req.get_method())
--> 519 response = self._open(req, data)
    521 # post-process response
    522 meth_name = protocol+"_response"

File /opt/conda/lib/python3.11/urllib/request.py:536, in OpenerDirector._open(self, req, data)
    533     return result
    535 protocol = req.type
--> 536 result = self._call_chain(self.handle_open, protocol, protocol +
    537                           '_open', req)
    538 if result:
    539     return result

File /opt/conda/lib/python3.11/urllib/request.py:496, in OpenerDirector._call_chain(self, chain, kind, meth_name, *args)
    494 for handler in handlers:
    495     func = getattr(handler, meth_name)
--> 496     result = func(*args)
    497     if result is not None:
    498         return result

File /opt/conda/lib/python3.11/urllib/request.py:1377, in HTTPHandler.http_open(self, req)
   1376 def http_open(self, req):
-> 1377     return self.do_open(http.client.HTTPConnection, req)

File /opt/conda/lib/python3.11/urllib/request.py:1352, in AbstractHTTPHandler.do_open(self, http_class, req, **http_conn_args)
   1350     except OSError as err: # timeout error
   1351         raise URLError(err)
-> 1352     r = h.getresponse()
   1353 except:
   1354     h.close()

File /opt/conda/lib/python3.11/http/client.py:1378, in HTTPConnection.getresponse(self)
   1376 try:
   1377     try:
-> 1378         response.begin()
   1379     except ConnectionError:
   1380         self.close()

File /opt/conda/lib/python3.11/http/client.py:318, in HTTPResponse.begin(self)
    316 # read until we get a non-100 response
    317 while True:
--> 318     version, status, reason = self._read_status()
    319     if status != CONTINUE:
    320         break

File /opt/conda/lib/python3.11/http/client.py:279, in HTTPResponse._read_status(self)
    278 def _read_status(self):
--> 279     line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
    280     if len(line) > _MAXLINE:
    281         raise LineTooLong("status line")

File /opt/conda/lib/python3.11/socket.py:706, in SocketIO.readinto(self, b)
    704 while True:
    705     try:
--> 706         return self._sock.recv_into(b)
    707     except timeout:
    708         self._timeout_occurred = True

TimeoutError: timed out
```

orenccl commented:
Hi @danhuawang,

Sorry about this. I found that the commit [Unified port number for Docker and Kubernetes](https://github.com/apache/gravitino-playground/pull/103/commits/72d532bd7d2c5102c8669dc45f0eb4568561aaf5) actually caused some issues with internal communication within the Kubernetes cluster.

This should be addressed in a separate issue, so I have reverted it and focused on the original problem. I've also updated the README to clarify the container and host port mapping, so users won't be confused about which port to forward.

Could you review this when you have free time?

![image](https://github.com/user-attachments/assets/76215019-25bd-422c-a7f4-d85595133826)

danhuawang commented on playground.sh:
```diff
@@ -18,186 +18,398 @@
   else
-    echo "ERROR: There was an issue running kubectl cluster-info, please check you K8s cluster."
+    # Check available space in the root partition if the directory doesn't exist (special case for WSL)
+    availableSpaceKB=$(df --output=avail / | awk 'NR==2 {print $1}')
```
I  encounter this issue when I launch with docker runtime. (MacOS)
```
df: unrecognized option `--output=avail'
usage: df [--libxo] [-b | -g | -H | -h | -k | -m | -P] [-acIilnY] [-,] [-T type] [-t type]
          [file | filesystem ...]
./playground.sh: line 90: / 1024 / 1024: syntax error: operand expected (error token is "/ 1024 / 1024")

```

orenccl replied:
It seems that macOS does not support `--output=avail`. Refer to [https://ss64.com/mac/df.html](https://ss64.com/mac/df.html).

Instead of using `--output=avail`, change it to `-k` so that it works on both macOS and Linux. Refer to [https://linuxcommand.org/lc3_man_pages/df1.html](https://linuxcommand.org/lc3_man_pages/df1.html).

However, I don't have a macOS environment. Could you check this for me?

danhuawang replied:
I checked macOS ,it's ok now.

danhuawang commented on playground.sh:
```diff
@@ -18,186 +18,398 @@
+
+  case "${runtime}" in
   k8s)
     kubectl -n gravitino-playground get pods -o wide
```
It's better to replace hard code namespace with ${playgroundRuntimeName} like the helm install command in start method

orenccl replied:
fixed

danhuawang commented:
@orenccl conflicts need to be resolved.

danhuawang commented:
@qqqttt123  Any comments?

xunliu commented:
hi @orenccl 
We plan to remove helm support, https://github.com/apache/gravitino-playground/pull/110
So, maybe this PR needs to wait for review.

orenccl commented:
@xunliu 
Sure, no problem.

orenccl commented:
I will update this PR soon since #110 has already been merged.

orenccl commented:
@xunliu @danhuawang 
Ready for review now! Please take a look when you have time

danhuawang commented:
LGTM. Thanks for your work!

xunliu reviewed (approved):
LGTM
